languages/cpp/reference/README.md

```diff
@@ -0,0 +1,158 @@
+# C++ reference
+
+## Concepts
+
+The C++ concept exercises are based on concepts. The list below contains the concepts that have been identified for the C++ language.
+
+### Object-oriented
+
+The core OO features a C++ developer should know about are:
+
+- Initialization
+- [Encapsulation](../../../reference/concepts/encapsulation.md)
+- [Classes](../../../reference/concepts/classes.md)
+- [Objects](../../../reference/concepts/objects.md)
+- [State](../../../reference/concepts/state.md)
+- [Mutation](../../../reference/concepts/mutation.md)
+- [Composition](../../../reference/concepts/composition.md)
+- [Inheritance](../../../reference/concepts/inheritance.md)
+- [Interfaces](../../../reference/concepts/interfaces.md)
+- [Runtime polymorphism](../../../reference/concepts/polymorphism.md)
+- [Methods](../../../reference/concepts/methods.md)
+
+### Functional
+
+C++ is a very flexible language, so lots of functional concepts apply to it:
+
+- [Immutability](../../../reference/concepts/immutability.md) (`const`ness)
+- [Higher-order functions](../../../reference/concepts/higher_order_functions.md)
+- [Type inference](../../../reference/concepts/type_inference.md) (`auto`, FTAD, and CTAD)
+- [Anonymous functions (lambdas)](../../../reference/concepts/anonymous_functions.md)
+- [Recursion](../../../reference/concepts/recursion.md)
+- [Nested functions](../../../reference/concepts/nested_functions.md)
+- [Pipelines (ranges)](../../../reference/concepts/pipelines.md) (Defer C++2a to release) 
+
+### Memory management
+
+- Object lifetime
+- RAII
+- Move-semantics
+
+### Functions
+
+- Function overloading
+- Default parameters
+- Special member-functions
+- Operators
+
+### General
+
+- Arithmetic
+- [Sameness](../../../reference/concepts/sameness.md) (Equality)
+- Ordering
+- [Conditionals](../../../reference/concepts/conditionals.md)
+- [Enumeration](../../../reference/concepts/enumeration.md) (Container iteration)
+- Iterators
+- Namespaces
+- IO
+- Templates
+- [Metaprogramming (TMP)](../../../reference/concepts/metaprogramming.md)
+- Exception handling
+- Implicit/explicit/contextual conversion
+- [Destructuring (structured-bindings)](../../../reference/concepts/destructuring_assignment.md)
+- Concurrency (thread-safety)
+- Expressions vs statements
+- Attributes
+- Slicing
+- Reflection (`<type_traits>`)
+- Undefined Behavior
+- Preprocessor
+- Concepts (Defer to C++2a)
+- Formatting (Defer to C++2a)
+
+### Types
+
+- Signed integers
+- Unsigned integers
+- Floating point numbers
+- [Characters][char]
+- [Booleans][bool]
+- Enums
+- [Strings][string]
+- [Arrays][array]
+- [Vectors][vector]
+- [Views][view] (`std::span` and `std::string_view`)
+- [Reference types][reference]
+- [Linked-lists][list]
+- [Dictionaries][map]
+- [Sets][set]
+- [Queues][queue]
+- [Stacks][stack]
+- [Tuples][tuple]
+- Unions (`std::variant`)
+- [Ranges][range] (Defer to C++2a)
+
+Let pointers aside as long as we can avoid them, using references instead.
```

Comment:
I don't think is a good idea to avoid raw pointers, they're a fundamental type of the language.

Reply:
We could simply classify them as view types, so with `std::span` and `std::string_view`

Reply:
Wouldn't this be misleading though? They're more than a simple view type.

Reply:
As in reading legacy code, yes
As in writing modern idiomatic C++, no; Bjarne himself said that `std::observer_ptr<T>` should not exist since `T*` should have that meaning already

Reply:
The only pointer property where I agree with you that views don't have, is arithmetic, because it can be a _LegacyRandomAccessIterator_, but I'm pretty sure everyone agrees that pointer arithmetic is a primary source of bugs, and that iterators types should just be `auto` or at worst `Container::iterator`

Reply:
Yeah, I like your point. Ditching them for now might be a good idea after all. 
Pointer arithmetic could be its own Concept maybe, but let's think about it later.

Reply:
I do think a concept on Pointer arithmetic would be useful. It might be a breeding ground for bugs but it's also one of the main ways to eke out performance from Cpp.

Reply:
As an advanced topic then; better not let the beginners breed bugs early